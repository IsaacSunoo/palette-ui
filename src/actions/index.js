export const setError = error => ({
  type: "SET_ERROR",
  error
});

export const setLoading = loading => ({
  type: "SET_LOADING",
  loading
});

export const setPalette = palette => ({
  type: "SET_PALETTE",
  palette
});

export const setLock = colorID => ({
  type: "SET_LOCK",
  colorID
});

export const setModal = status => ({
  type: "SET_MODAL",
  status
});

export const setProjects = projects => ({
  type: "SET_PROJECTS",
  projects
});
