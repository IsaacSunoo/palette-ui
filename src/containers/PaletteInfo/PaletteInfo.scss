.PaletteInfo {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  margin-bottom: .6em;

  h4 {
    margin: 0;
    font-family: $fontRobotoCond;
    width: 45%;
    padding-left: .4em;
    padding-right: .5em;
    color: #ffffff29;
    font-weight: 100;
  }

  ul {
    display: flex;
    padding: 0;
    margin: 0;
    width: 70%;
    height: 100%;
    padding-right: .5em;

    li {
      list-style-type: none;
      width: 20%;
    }
  }

  .PaletteInfo-Delete {

  }

  .DeleteBtn {
    border: none;
    background: none;
    outline: none;
    border-radius: 5px;
    padding: .3em .6em;
    cursor: pointer;
    color: #ffffff29;
    transition: all .5s ease;
    opacity: 1;

    &:hover {
      opacity: .5;
    }

    i {
      font-weight: 100;
      font-size: 1.1em;
    }
  }

}
