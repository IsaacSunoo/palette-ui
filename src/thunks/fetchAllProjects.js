import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { cleanProjectsPalettes } from '../utility/cleaner';

export const fetchAllProjects = () => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const allProjects = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/projects');
      const allPalettes = await fetchData(process.env.REACT_APP_BACKEND_URL + '/api/v1/palettes');
      dispatch(actions.setLoading(false));
      if(!allProjects.ok){
        throw Error(allProjects.statusText)
      }
      const combinedData = cleanProjectsPalettes(allProjects, allPalettes)
      dispatch(actions.setProjects(combinedData));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
