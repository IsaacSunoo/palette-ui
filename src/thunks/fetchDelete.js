import * as actions from '../actions/index';
import { fetchData } from '../utility/fetchData';
import { fetchOptions } from '../utility/fetchOptions';

export const fetchDelete = (url) => {
  return async (dispatch) => {
    try {
      dispatch(actions.setLoading(true));
      const options = await fetchOptions('DELETE', {});
      await fetchData(url, options);
      dispatch(actions.setLoading(false));
    } catch (error) {
      dispatch(actions.setError(error.message));
    }
  }
}
