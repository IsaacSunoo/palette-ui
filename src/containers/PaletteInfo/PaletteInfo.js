import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { fetchAllProjects } from '../../thunks/fetchAllProjects';
import { fetchDelete } from '../../thunks/fetchDelete';
import * as actions from '../../actions/index';

export class PaletteInfo extends Component {

  handleDelete = async (id) => {
    const { fetchDelete, fetchAllProjects } = this.props;
    const url = `http://localhost:3001/api/v1/palettes/${id}`;
    await fetchDelete(url);
    fetchAllProjects();
  }

  handleGenerate = () => {
    const { color_1, color_2, color_3, color_4, color_5, setPalette } = this.props;
    let palette = [
      { color: color_1, locked: true },
      { color: color_2, locked: true },
      { color: color_3, locked: true },
      { color: color_4, locked: true },
      { color: color_5, locked: true }
    ]
    setPalette(palette);
  }

  render() {
    const { name, id, color_1, color_2, color_3, color_4, color_5 } = this.props;

    return (
      <div className='PaletteInfo'>
        <h4>{name}</h4>
        <ul>
          <li style={{backgroundColor: color_1}}>&nbsp;</li>
          <li style={{backgroundColor: color_2}}>&nbsp;</li>
          <li style={{backgroundColor: color_3}}>&nbsp;</li>
          <li style={{backgroundColor: color_4}}>&nbsp;</li>
          <li style={{backgroundColor: color_5}}>&nbsp;</li>
        </ul>
        <section className='PaletteInfo-View'>
          <button className='ShowBtn' onClick={this.handleGenerate}>
            <i className='fas fa-eye'></i>
          </button>
        </section>
        <section className='PaletteInfo-Delete'>
          <button className='DeleteBtn' onClick={() => this.handleDelete(id)}>
            <i className='fas fa-trash-alt'></i>
          </button>
        </section>
      </div>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  setPalette: data => dispatch(actions.setPalette(data)),
  fetchAllProjects: data => dispatch(fetchAllProjects(data)),
  fetchDelete: data => dispatch(fetchDelete(data)),
});

PaletteInfo.propTypes = {
  setPalette: PropTypes.func,
  fetchAllProjects: PropTypes.func,
  fetchDelete: PropTypes.func,
}

export default connect(null, mapDispatchToProps)(PaletteInfo);
