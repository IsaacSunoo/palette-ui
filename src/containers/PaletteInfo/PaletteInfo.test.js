import React from 'react';
import { shallow } from 'enzyme'
import { PaletteInfo, mapDispatchToProps } from './PaletteInfo';
import * as actions from '../../actions/index';
import { fetchAllProjects } from '../../thunks/fetchAllProjects';
import { fetchDelete } from '../../thunks/fetchDelete';
jest.mock('../../thunks/fetchAllProjects');
jest.mock('../../thunks/fetchDelete');

const mockPalette = {
  id: 15,
  name: "Mikes Pal",
  color_1: "#93125a",
  color_2: "#ffffff",
  color_3: "#f13470",
  color_4: "#80064e",
  color_5: "#423c52",
  project_id: 8,
}

describe('PaletteInfo', () => {

  describe('PaletteInfo Component', () => {
    let wrapper;
    let mockfetchAllProjects = jest.fn()
    let mockSetPalette = jest.fn()
    let mockFetchDelete = jest.fn();


    beforeEach(() => {
      wrapper = shallow(<PaletteInfo fetchAllProjects={mockfetchAllProjects}
                                     setPalette={mockSetPalette}
                                     fetchDelete={mockFetchDelete}
                                     color_1="#93125a"
                                     color_2="#ffffff"
                                     color_3="#f13470"
                                     color_4="#80064e"
                                     color_5="#423c52"
                                     id={15}
                                     project_id={8}
                      />)
    })

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should call handleGenerate when ShowButton is clicked', () => {
      wrapper.instance().handleGenerate = jest.fn()
      wrapper.instance().forceUpdate()
      wrapper.update()
      wrapper.find('.ShowBtn').simulate('click')
      expect(wrapper.instance().handleGenerate).toHaveBeenCalled()
    });

    it('should call setPalette when handleGenerate is invoked with mockPalette', () => {
      const instance = wrapper.instance();
      const mockCurrentPalette = [
        { color: mockPalette.color_1, locked: true },
        { color: mockPalette.color_2, locked: true },
        { color: mockPalette.color_3, locked: true },
        { color: mockPalette.color_4, locked: true },
        { color: mockPalette.color_5, locked: true }
      ]
      instance.handleGenerate();
      expect(mockSetPalette).toHaveBeenCalledWith(mockCurrentPalette);
    });

    it('should call fetchAllProjects when handleDelete is invoked with id', async () => {
      const instance = wrapper.instance();
      const mockId = 8;
      await instance.handleDelete(mockId);
      expect(mockfetchAllProjects).toHaveBeenCalled();
    });

    it('should call handleDelete on click with the correct id passed in', () => {
      const mockId = 15;
      const instance = wrapper.instance();
      const spy = jest.spyOn(instance, 'handleDelete');
      const button = wrapper.find('.DeleteBtn');
      button.simulate('click', mockId);
      expect(spy).toHaveBeenCalledWith(mockId);
    });

    it('should call fetchDelete when handDelete is invoked', async () => {
      const instance = wrapper.instance();
      const mockId = 15;
      await instance.handleDelete(mockId);
      expect(mockFetchDelete).toHaveBeenCalled();
    });

    it('should call fetchAllProjects when handDelete is invoked', async () => {
      const instance = wrapper.instance();
      const mockId = 15;
      await instance.handleDelete(mockId);
      expect(mockfetchAllProjects).toHaveBeenCalled();
    });

  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch for setPalette', () => {
      const mockData = []
      const mockDispatch = jest.fn();
      const actionToDispatch = actions.setPalette(mockData)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.setPalette(mockData)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch for fetchAllProjects', () => {
      const mockData = []
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchAllProjects(mockData)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.fetchAllProjects(mockData)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call dispatch for fetchDelete', () => {
      const mockData = []
      const mockDispatch = jest.fn();
      const actionToDispatch = fetchDelete(mockData)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.fetchDelete(mockData)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });

  });

});
