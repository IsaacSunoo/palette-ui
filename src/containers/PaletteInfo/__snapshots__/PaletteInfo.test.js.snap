// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaletteInfo PaletteInfo Component should match the snapshot 1`] = `
<div
  className="PaletteInfo"
>
  <h4 />
  <ul>
    <li
      style={
        Object {
          "backgroundColor": "#93125a",
        }
      }
    >
       
    </li>
    <li
      style={
        Object {
          "backgroundColor": "#ffffff",
        }
      }
    >
       
    </li>
    <li
      style={
        Object {
          "backgroundColor": "#f13470",
        }
      }
    >
       
    </li>
    <li
      style={
        Object {
          "backgroundColor": "#80064e",
        }
      }
    >
       
    </li>
    <li
      style={
        Object {
          "backgroundColor": "#423c52",
        }
      }
    >
       
    </li>
  </ul>
  <section
    className="PaletteInfo-View"
  >
    <button
      className="ShowBtn"
      onClick={[Function]}
    >
      <i
        className="fas fa-eye"
      />
    </button>
  </section>
  <section
    className="PaletteInfo-Delete"
  >
    <button
      className="DeleteBtn"
      onClick={[Function]}
    >
      <i
        className="fas fa-trash-alt"
      />
    </button>
  </section>
</div>
`;
